packages/block-editor/src/hooks/fit-text.js

```diff
@@ -0,0 +1,262 @@
+			return;
+		}
+
+		// Get or create style element with unique ID
+		const styleId = `fit-text-${ clientId }`;
+		let styleElement = blockElement.ownerDocument.getElementById( styleId );
+		if ( ! styleElement ) {
+			styleElement = blockElement.ownerDocument.createElement( 'style' );
+			styleElement.id = styleId;
+			blockElement.ownerDocument.head.appendChild( styleElement );
+		}
+
+		const blockSelector = `#block-${ clientId }`;
+
+		const applyStylesFn = ( css ) => {
+			styleElement.textContent = css;
+		};
+
+		optimizeFitText( blockElement, blockSelector, applyStylesFn );
+	}, [ blockElement, clientId, hasFitTextSupport, fitText ] );
+
+	useEffect( () => {
+		if (
+			! fitText ||
+			! blockElement ||
+			! clientId ||
+			! hasFitTextSupport
+		) {
+			return;
+		}
+
+		// Apply initially
+		applyFitText();
+
+		// Store current element value for cleanup
+		const currentElement = blockElement;
+
+		// Watch for size changes
+		let resizeObserver;
+		if ( window.ResizeObserver && currentElement.parentElement ) {
+			resizeObserver = new window.ResizeObserver( applyFitText );
+			resizeObserver.observe( currentElement.parentElement );
+		}
+
+		// Cleanup function
+		return () => {
+			if ( resizeObserver ) {
+				resizeObserver.disconnect();
+			}
+
+			const styleId = `fit-text-${ clientId }`;
+			const styleElement =
+				currentElement.ownerDocument.getElementById( styleId );
+			if ( styleElement ) {
+				styleElement.remove();
+			}
+		};
+	}, [ fitText, clientId, applyFitText, blockElement, hasFitTextSupport ] );
+
+	// Trigger fit text recalculation when content changes
+	useEffect( () => {
+		if ( fitText && blockElement && hasFitTextSupport ) {
+			// Small delay to ensure DOM has updated after content changes
+			const timer = setTimeout( () => {
+				if ( blockElement ) {
+					applyFitText();
+				}
+			}, 100 );
+
+			return () => clearTimeout( timer );
+		}
+	}, [
+		blockAttributes,
+		fitText,
+		applyFitText,
+		blockElement,
+		hasFitTextSupport,
+	] );
+}
+
+/**
+ * Fit text control component for the typography panel.
+ *
+ * @param {Object}   props               Component props.
+ * @param {string}   props.clientId      Block client ID.
+ * @param {Function} props.setAttributes Function to set block attributes.
+ * @param {string}   props.name          Block name.
+ * @param {boolean}  props.fitText       Whether fit text is enabled.
+ */
+export function FitTextControl( { clientId, fitText, setAttributes, name } ) {
+	if ( ! hasBlockSupport( name, FIT_TEXT_SUPPORT_KEY ) ) {
+		return null;
+	}
+	return (
+		<InspectorControls group="typography">
+			<ToolsPanelItem
+				hasValue={ () => fitText }
+				label={ __( 'Fit text' ) }
+				onDeselect={ () => setAttributes( { fitText: false } ) }
+				resetAllFilter={ () => ( { fitText: false } ) }
+				panelId={ clientId }
+			>
+				<ToggleControl
+					__nextHasNoMarginBottom
+					label={ __( 'Fit text' ) }
+					checked={ fitText }
```

Comment:
```suggestion
					checked={ !! fitText }
```

This can be undefined and causes the error @scruffian [mentioned](https://github.com/WordPress/gutenberg/pull/71904#issuecomment-3365236372).

Reply:
Ah, I wrongly assumed there was a default value set up.